src/Cake.Common/IO/FileCopier.cs

```diff
@@ -103,10 +105,28 @@ public static void CopyFiles(ICakeContext context, IEnumerable<FilePath> filePat
                 throw new DirectoryNotFoundException(message);
             }
 
-            // Iterate all files and copy them.
-            foreach (var filePath in filePaths)
+            if (preserverFolderStructure)
             {
-                CopyFileCore(context, filePath, absoluteTargetDirectoryPath.GetFilePath(filePath));
+                var matches = from len in Enumerable.Range(0, filePaths.Min(s => s.ToString().Length)).Reverse() ////http://stackoverflow.com/a/8578268/2845744
```

Comment:
I would rather see this as imperative code instead of LINQ since the matching is very error prone in it's current state and difficult to reason about.